### PR TITLE
[KYLIN-4382] Unable to use DATE type in prepared statements

### PR DESCRIPTION
**Pull request by juntaozhang**
## Proposed changes

fix[ KYLIN-4382](https://issues.apache.org/jira/browse/KYLIN-4382) Unable to use DATE type in prepared statements

## Github Branch 

As most of the development works are on Kylin 4, we need to switch it as main branch. Apache Kylin community changes the branch settings on Github since 2021-08-04 :

1. The original branch _kylin-on-parquet-v2_ for **Kylin 4.X** (Parquet Storage) has been renamed to branch **main**, and configured as the **default** branch;
2. The original branch _master_ for **Kylin 3.X** (HBase Storage) has been renamed to branch **kylin3** ;

Please check [Intro to Kylin 4 architecture](https://kylin.apache.org/blog/2021/07/02/Apache-Kylin4-A-new-storage-and-compute-architecture/) and [INFRA-22166](https://issues.apache.org/jira/browse/INFRA-22166) if you are interested.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [x] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: query/src/main/java/org/apache/kylin/query/relnode/OLAPContext.java [hunk 1]
```diff
@@ -337,7 +340,14 @@ private String transferDateTimeColumnToMillis(CompareTupleFilter compFilter, Str
...
         if (column.getType().isDateTimeFamily()){
-            value = String.valueOf(DateFormat.stringToMillis(value));
+            logger.debug("value changed before: {}", value);
+            if(column.getType().isDate()) {
+                // It seems the dynamic parameter has been changed to the date integer value
+                value = String.valueOf(Long.parseLong(value) * 86400000L);
+            } else {
+                value = String.valueOf(DateFormat.stringToMillis(value));
+            }
+            logger.debug("value changed after: {}", value);
```
Comment (hit-lacus) [inline query/src/main/java/org/apache/kylin/query/relnode/OLAPContext.java:350]: Could you please combine two lines into one line, like : `Column value changed from {} to {}.`

**Comment by zhangayqian** [comment]
Hi @juntaozhang, I tested this PR, which can fix the bug on kylin3, but it doesn't work on kylin4. I think this PR should be merged into `kylin3` branch.

File: jdbc/src/test/java/org/apache/kylin/jdbc/DriverTest.java [hunk 2]
```diff
@@ -286,7 +286,7 @@ public void testKYLIN4382() throws SQLException, ParseException {
         info.put("password", "KYLIN");
         Connection conn = driver.connect("jdbc:kylin://localhost:7070/default", info);
         PreparedStatement state = conn.prepareStatement("select count(*) from test_kylin_fact where cal_dt=?");
-        state.setDate(1, new Date(new SimpleDateFormat("yyyy-MM-dd").parse("2012-01-01").getTime()));
+        state.setDate(1, new Date(FastDateFormat.getInstance("yyyy-MM-dd").parse("2012-01-01").getTime()));
```
Comment (zhangayqian) [inline jdbc/src/test/java/org/apache/kylin/jdbc/DriverTest.java:289]: I can't understand why use `FastDateFormat` to replace `SimpleDateFormat`.

**Comment by juntaozhang** [comment]
> Hi @juntaozhang, I tested this PR, which can fix the bug on kylin3, but it doesn't work on kylin4. I think this PR should be merged into `kylin3` branch.

Thanks for your reply, do you means I should merge to branch/kylin3?  Need I create a new PR?

**Comment by zhangayqian** [comment]
> > Hi @juntaozhang, I tested this PR, which can fix the bug on kylin3, but it doesn't work on kylin4. I think this PR should be merged into `kylin3` branch.
> 
> Thanks for your reply, do you means I should merge to branch/kylin3? Need I create a new PR?

Yes, you can create a new PR to `kylin3`.

**Pull request closed without merging**